Userland/Libraries/LibSemVer/SemVer.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) 2023, Gurkirat Singh <tbhaxor@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <AK/Find.h>
+#include <AK/Function.h>
+#include <AK/GenericLexer.h>
+#include <AK/ReverseIterator.h>
+#include <AK/StringBuilder.h>
+#include <LibRegex/Regex.h>
+#include <LibRegex/RegexParser.h>
+#include <LibSemVer/SemVer.h>
+
+namespace SemVer {
+String SemVer::suffix() const
+{
+    StringBuilder sb;
+    if (m_prerelease_identifiers.size() > 0)
+        sb.appendff("-{}", prerelease());
+    if (m_buildmetadata.code_points().length() > 0)
+        sb.appendff("+{}", buildmetadata());
+    return MUST(sb.to_string());
+}
+
+String SemVer::to_string() const
+{
+    return MUST(String::formatted("{}{}{}{}{}{}", m_major, m_number_separator, m_minor, m_number_separator, m_patch, suffix()));
+}
+
+ErrorOr<SemVer> SemVer::bump(BumpType type) const
+{
+    switch (type) {
+    case BumpType::Major:
+        return SemVer(m_major + 1, 0, 0, m_number_separator);
+    case BumpType::Minor:
+        return SemVer(m_major, m_minor + 1, 0, m_number_separator);
+    case BumpType::Patch:
+        return SemVer(m_major, m_minor, m_patch + 1, m_number_separator);
+    case BumpType::Prerelease: {
+        Vector<String> prerelease_identifiers = m_prerelease_identifiers;
+        bool is_found = false;
+
+        // Unlike comparision, prerelease bumps take from RTL
+        for (auto& identifier : AK::ReverseWrapper::in_reverse(prerelease_identifiers)) {
+            auto numeric_identifier = identifier.to_number<u32>();
+            if (numeric_identifier.has_value()) {
+                is_found = true;
+                identifier = TRY(String::formatted("{}", numeric_identifier.value() + 1));
+                break;
+            }
+        }
+
+        // Append 0 identifier if there is no numeric found to be bumped
+        if (!is_found)
+            prerelease_identifiers.append("0"_string);
+
+        return SemVer(m_major, m_minor, m_patch, m_number_separator, prerelease_identifiers, ""_string);
+    }
+    default:
+        return Error::from_string_view("Invalid bump type"sv);
```

Comment:
If there is no normal way that `bump` returns an error, you should leave the `ErrorOr<SemVer>` out of the API and instead just return `SemVer` and replace the line above with `VERIFY_NOT_REACHED()`. This makes the code simpler for callers.

Reply:
> instead just return SemVer

Based on my experience, the patch version is more likely to change as it reflects ongoing maintenance, bug fixing, and security updates. We should add it as `default`.

If neither of major or minor, then patch update.

Reply:
I'm not sure what you're talking about - my point is that instead of `ErrorOr<SemVer>`, you could return `SemVer` to make things a bit easier to use.

Reply:
Another way to return `Optional<SemVer>` but, that doesn't make sense to me.

Reply:
Why would it be `Optional<SemVer>`? The current code always seems to return a `SemVer`, except when there's an invalid `BumpType`. That case you could handle with `VERIFY_NOT_REACHED()` imho.

Reply:
> I'm not sure what you're talking about

My point was to not only return any kind of error, not even the one thrown by VERIFY_NOT_REACHED. But using it regardless now, seems to be more obvious than what I had in my mind earlier.

Reply:
VERIFY_NOT_REACHED does not throw an error, or return an error. it crashes the program because there is a precondition violiation.